return table.freeze {
	null = require "@self/primitive/null",
	boolean = require "@self/primitive/boolean",
	string = require "@self/primitive/string",
	csetstring = require "@self/primitive/csetstring",
	buffer = require "@self/primitive/buffer",
	number = require "@self/primitive/number",
	vector = require "@self/primitive/vector",
	array = require "@self/primitive/array",
	bitarray = require "@self/primitive/bitarray",
	struct = require "@self/primitive/struct",
	set = require "@self/primitive/set",
	union = require "@self/primitive/union",

	CFrame = require "@self/roblox/CFrame",
	Enum = require "@self/roblox/Enum",
	Vector2 = require "@self/roblox/Vector2",
	Vector3 = require "@self/roblox/Vector3",
	BrickColor = require "@self/roblox/BrickColor",
	Color3 = require "@self/roblox/Color3",
}
