local okserde = require("./okserde")

return okserde({
	boolean = require("./schemas/language/boolean"),
	buffer = require("./schemas/language/buffer"),
	null = require("./schemas/language/null"),
	number = require("./schemas/language/number"),
	string = require("./schemas/language/string"),
	vector = require("./schemas/language/vector"),

	array = require("./schemas/compound/array"),
	optional = require("./schemas/compound/optional"),
	struct = require("./schemas/compound/struct"),
	tuple = require("./schemas/compound/tuple"),
	union = require("./schemas/compound/union"),

	bitarray = require("./schemas/extra/bitarray"),
	bitstruct = require("./schemas/extra/bitstruct"),
	charsetstring = require("./schemas/extra/charsetstring"),
	flags = require("./schemas/extra/flags"),
	nullstring = require("./schemas/extra/nullstring"),
	set = require("./schemas/extra/set"),
	termstring = require("./schemas/extra/termstring"),
	void = require("./schemas/extra/void"),

	Axes = require("./schemas/roblox/Axes"),
	BrickColor = require("./schemas/roblox/BrickColor"),
	CFrame = require("./schemas/roblox/CFrame"),
	Color3 = require("./schemas/roblox/Color3"),
	ColorSequence = require("./schemas/roblox/ColorSequence"),
	ColorSequenceKeypoint = require("./schemas/roblox/ColorSequenceKeypoint"),
	Content = require("./schemas/roblox/Content"),
	DateTime = require("./schemas/roblox/DateTime"),
	Enum = require("./schemas/roblox/Enum"),
	EnumItem = require("./schemas/roblox/EnumItem"),
	Faces = require("./schemas/roblox/Faces"),
	Font = require("./schemas/roblox/Font"),
	NumberRange = require("./schemas/roblox/NumberRange"),
	NumberSequence = require("./schemas/roblox/NumberSequence"),
	NumberSequenceKeypoint = require("./schemas/roblox/NumberSequenceKeypoint"),
	PathWaypoint = require("./schemas/roblox/PathWaypoint"),
	PhysicalProperties = require("./schemas/roblox/PhysicalProperties"),
	Ray = require("./schemas/roblox/Ray"),
	Rect = require("./schemas/roblox/Rect"),
	Region3 = require("./schemas/roblox/Region3"),
	TweenInfo = require("./schemas/roblox/TweenInfo"),
	UDim = require("./schemas/roblox/UDim"),
	UDim2 = require("./schemas/roblox/UDim2"),
	Vector2 = require("./schemas/roblox/Vector2"),
	Vector3 = require("./schemas/roblox/Vector3"),
})
